### DISCUSSION TITLE
Reconnect option

### DISCUSSION THREAD

**Discussion by ruicanas** [adam-fowler/mqtt-nio#140] [Ideas]
Hey guys! I'm currently using this library and I saw that, when the client fails it first connection, there is no chance of actually trying to reconnect automatically. My request would be: have a reconnect option that could be activated in the MQTT client's configuration. Would that be possible to do?

**Comment by adam-fowler** [comment]
Could you not just write
```swift
while true {
    do {
        try await client.connect()
        break
    } catch MQTTError.failedToConnect {
        try await Task.sleep(for: someExponentialIncreasingTimeAmount)
    }
}
```

**Reply by ruicanas** [reply]
Sure, that would work. I was thinking in something related to the client configuration, because it would be more practical (in my opinion, of course) to have it there since it ends up being an actual configuration - having an auto reconnect (with the logic on in the library) or not.
If it doesn't make sense, I can close this post 🙂

**Reply by adam-fowler** [reply]
There are quite a few parameters to consider when doing a retry system. 
- how long to wait between retries
- Should wait time increase exponentially/linearly after each fail and by how much
- Should wait time increase be jittered
- Should it fail after so many retries
- Should it fail after so much time
I would rather the individual wrote their own, instead of the client providing an implementation that tries to support every combination of the above

**Reply by ruicanas** [reply]
Okay, those are very fair points and it ends up being good since it give us some freedom to build a reconnect mechanism. However, there is a point that you didn't mention that was the re-subscription to the topics that were subscribed before. It makes sense to me that we need to subscribe to our topics again, but this mechanism in the library side would be awesome to take over this matter.

**Reply by adam-fowler** [reply]
Resubscription should work, if you pass in `cleanSession: false` when calling `connect`

**Reply by ruicanas** [reply]
Hey, sorry for the delay! 
Yes, it should but it seems like the connection is being made but the subscriptions are all lost (even with the ``cleanSession: false``), since I stop receiving info from the broker ☹️
<details>
<summary>Implementation 🪄</summary>

```swift
func addReconnectListener() {
        
    guard let client else {
        log.error("\(Tags.mqtt) \(Strings.notInitialized)")
        return
    }
    
    client.removeCloseListener(named: Tags.reconnect)
    client.addCloseListener(named: Tags.reconnect) { _ in
        log.warning("\(Tags.mqtt) Connection with MQTT was closed!")
        Task { await self.reconnect() }
    }
}
```

```swift
func reconnect() async {
        
    guard let client, !client.isActive() else { return }
    log.debug("Reconnecting…")
    
    do {
        _ = try await client.connect(cleanSession: false).get()
        log.debug("Connection successful")
    } catch {
        log.debug("Error while connecting: \(error)")
        try? await Task.sleep(nanoseconds: Time.twoSecondsInNanoseconds)
        await reconnect()
    }
}
```

</details>

**Reply by adam-fowler** [reply]
Can you show the code where you are doing the subscription and how you are reading the publish messages.

**Reply by ruicanas** [reply]
Sure, here it goes:

<details>
<summary>Subscribe</summary>

```swift
func subscribe(
    topic: String,
    qualityOfService: MqttQosLevel
) async {
    
    guard let client else {
        log.error("\(Tags.mqtt) \(Strings.notInitialized)")
        return
    }
    
    let subs = MQTTSubscribeInfo(
        topicFilter: topic,
        qos: qualityOfService.asMQTTQoS
    )
    
    do {
        _ = try await client.subscribe(to: [subs]).get()
    } catch {
        log.debug("\(Tags.mqtt) Subscribe with error. - \(error)")
    }
}
```

</details>

<details>
<summary>Reading Topics</summary>

```swift
func receiveTopicFlow(
    topic: String,
    messageConsumer: MqttMessageConsumer
) {
    
    guard let client else {
        log.error("\(Tags.mqtt) \(Strings.notInitialized)")
        return
    }
    
    client.addPublishListener(named: "\(Tags.publisher)\(topic)") { result in
        switch result {
        case .success(let packet):
            if topic == packet.topicName {
                var buffer = packet.payload
                let message = buffer.readString(length: buffer.readableBytes)
                
                messageConsumer.consumeMessage(message: message ?? Strings.noMessage) {
                    guard let error = $0 else { return }
                    log.error("\(Tags.mqtt) Consume message with error - \(error)")
                }
            }
        case .failure(let error):
            log.debug("\(Tags.mqtt) - Publish listener error - \(error)")
        }
    }
}
```
</details>

**Reply by adam-fowler** [reply]
There is no chance you are using a new client with a different identifier when reconnecting?

**Reply by ruicanas** [reply]
Nope, I keep using the same client that was using the same configs to connect the first time. I've made some logs to confirm that my client configurations were untouched at the moment the reconnecting was being done.

**Reply by adam-fowler** [reply]
Do you set cleanSession to false every time you connect? 

Just been going through the specification. You are required to set cleanSession to false for all sessions if you want to store sessions between connections. If you set it to true when that session closes it discards any state.

**Reply by ruicanas** [reply]
Nope, I was not aware that it was necessary to have that property always set on false. However, after I switched the ``cleanSession`` to false for all sessions, it keeps discarding the previous session ☹️

**Reply by adam-fowler** [reply]
From what I can tell v3.1 requires cleanSession set all the time, but v5 of the protocol doesn't. Which version of the protocol are you using? Actually if you could provide your full Configuration (minus any identifying bits) that might help as well.

**Reply by ruicanas** [reply]
Okay, I'm using v5 so that's probably not the problem then. Regarding my configuration, I'm also sending how I'm doing the connection as well:

<details>
<summary>Client Configuration</summary>

```swift
// SSL/TLS Configuration

let tlsConfig = TSTLSConfiguration(
    minimumTLSVersion: .tlsV12,
    trustRoots: awsCertificate,
    applicationProtocols: [Tags.MQTT.mqttProtocol]
)

// Client Configuration

clientConfig = MQTTClient.Configuration(
    version: .v5_0,
    keepAliveInterval: .seconds(Int64(keepAliveIntervalInSeconds)),
    connectTimeout: .seconds(Int64(sessionExpiryIntervalInSeconds)),
    timeout: .seconds(Int64(10)),
    userName: simpleAuth?.username,
    password: simpleAuth?.password,
    useSSL: true,
    tlsConfiguration: .ts(tlsConfig)
)
```
</details>

<details>
<summary>Connection</summary>

```swift

func connect() {
  let mqttClient: MQTTClient = .init(
      host: host,
      port: port,
      identifier: id,
      eventLoopGroupProvider: .createNew,
      configuration: clientConfig
  )
  
  await connect(client: mqttClient)
}

func connect(
    client: MQTTClient,
    cleanSession: Bool = true, // When doing the reconnection mechanism, I set this flag to false.
    will: LastWill? = nil
) async {
    
    var retries = 0
    while retries < Retries.connectRetries {
        do {
            _ = try await client.connect(
                cleanSession: cleanSession,
                will: will
            )
            .map({ _ in self.addReconnectListener() })
            .get()
            break
        } catch {
            log.debug("\(Tags.mqtt) Connect with error. - \(error)")
            try? await Task.sleep(nanoseconds: Time.oneSecondInNanoseconds)
            retries += 1
        }
    }
}
```
</details>

**Reply by adam-fowler** [reply]
You did try with setting this to false all the time? As the example code here defaults to setting it to true. 
```swift
func connect(
    client: MQTTClient,
    cleanSession: Bool = true, // When doing the reconnection mechanism, I set this flag to false.
    will: LastWill? = nil
) async {
```
More questions:
The `connect` function returns whether there was a session stored for the client. When you call `connect` the second time does it return true or false?
I have also updated the persistent session test https://github.com/swift-server-community/mqtt-nio/blob/acab3749ef1facf18aa25435dd30c246c374935b/Tests/MQTTNIOTests/MQTTNIOTests.swift#L428C19-L428C19 to verify persistent sessions are working. It sets up two clients and performs the following actions
- client1 connects
- client2 connects
- client2 subscribes to topic
- client1 publishes to topic
- client2 disconnects
- client2 connects
- client1 publishes to topic

If the session is kept between connections then client2 should receive both messages published to the topic

To run tests you need to run a local mosquitto server. There is a script `.scripts/mosquitto.sh` that'll use docker to run one for you.

**Reply by ruicanas** [reply]
Sorry for the delay, but more work got in the middle, and I wasn't able to set things up as fast as I would have liked to.

I was able to set up a local Mosquitto server, and I've run the test you mentioned (as well as other tests), and everything is working as expected.

About the first question: So... I forgot that default parameter. I ended up changing it to false, and I've also set the ``cleanSession`` to false in the ``connect`` inside my reconnect mechanism. My client was able to establish the connection again and receive information from all the previously subscribed topics.

About the second question: I was not handling that return value, but I've already written some code to handle it. If the connection was not established, it tries to connect again.

However, I still find this odd because since I'm configuring my client with a ``.V5``, the reconnect should work despite having my first connection's ``cleanSession`` set to true. I've also tried to do the ``client.v5.connect`` like it was being done in the tests (already handling the value returned from the connect), and the results are pretty much the same: it seems like it stops receiving information from the subscribed topics.

**Reply by adam-fowler** [reply]
Reading the v5 spec section on `cleanStart` does not make it clear whether a session is stored on connection close. It only mentions if a session exists. You also need to look at the section on the session expiry interval property https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901048

**Reply by ruicanas** [reply]
After revisiting the documentation and reading your explanation about the sessionExpiryInterval, something clicked in my head: the issue was that I hadn't set this property during my connection. Even though I had the flag set to 'false,' it didn't matter because the session would close immediately after the disconnect.

Is it possible to include sessionExpiryInterval in the MQTTClient.Configuration? I believe it would make things much more straightforward.

I also have another question: Is there any way to change client properties? I need to update my client's password from time to time, and it seems I can't do that without shutting down the client. However, when I use shutdown(), it appears that I can't recover my previous session.

**Reply by adam-fowler** [reply]
> Is it possible to include sessionExpiryInterval in the MQTTClient.Configuration? I believe it would make things much more straightforward.
> 
Not overly keen on this as it doesn't mean anything on v3.1.1 clients, but I also see what you mean. Let me have a think about this 

> I also have another question: Is there any way to change client properties? I need to update my client's password from time to time, and it seems I can't do that without shutting down the client. However, when I use shutdown(), it appears that I can't recover my previous session.

Perhaps I could add a way to override the configuration when you call `connect`. Would that work for you?

**Reply by ruicanas** [reply]
> Not overly keen on this as it doesn't mean anything on v3.1.1 clients, but I also see what you mean. Let me have a think about this

Well, it's just an idea but maybe the ``.v5`` client parameter could be used for that purpose.

> Perhaps I could add a way to override the configuration when you call connect. Would that work for you?

So, basically using ``connect`` where I could insert my new configurations as a parameter, right? Yap, that would work for me!

**Reply by adam-fowler** [reply]
> I also have another question: Is there any way to change client properties? I need to update my client's password from time to time, and it seems I can't do that without shutting down the client. However, when I use shutdown(), it appears that I can't recover my previous session.

Can you verify that the session state is lost because you used a new client and not because the connection credentials changed

**Reply by ruicanas** [reply]
So, I reused some of the code you had written in the unit tests and made some minor changes. The test below will fail with unfulfilled expectations:

<details>
<summary>Unit Test</summary>

```swift
    func testPersistentSession() throws {
        let expectation = XCTestExpectation(description: "testPersistentSession")
        expectation.expectedFulfillmentCount = 2
        expectation.assertForOverFulfill = true
        
        let payloadString = #"{"from":1000000,"to":1234567,"type":1,"content":"I am a beginner in swift and I am studying hard!!测试\n\n test, message","timestamp":1607243024,"nonce":"pAx2EsUuXrVuiIU3GGOGHNbUjzRRdT5b","sign":"ff902e31a6a5f5343d70a3a93ac9f946adf1caccab539c6f3a6"}"#
        let payload = ByteBufferAllocator().buffer(string: payloadString)
        
        let client = self.createClient(identifier: "testPersistentSession_publisher")
        defer { XCTAssertNoThrow(try client.syncShutdownGracefully()) }
        _ = try client.connect().wait()
        var client2 = self.createClient(identifier: "testPersistentSession_subscriber")
        defer { XCTAssertNoThrow(try client2.syncShutdownGracefully()) }
        
        client2.addPublishListener(named: "test") { result in
            switch result {
            case .success(let publish):
                var buffer = publish.payload
                let string = buffer.readString(length: buffer.readableBytes)
                XCTAssertEqual(string, payloadString)
                expectation.fulfill()
                
            case .failure(let error):
                XCTFail("\(error)")
            }
        }
        _ = try client2.v5.connect(cleanStart: true, properties: .init([.sessionExpiryInterval(60)])).wait()
        _ = try client2.subscribe(to: [.init(topicFilter: "testPersistentAtLeastOnce", qos: .atLeastOnce)]).wait()
        try client.publish(to: "testPersistentAtLeastOnce", payload: payload, qos: .atLeastOnce).wait()
        Thread.sleep(forTimeInterval: 1)
        try client2.disconnect().wait()
        try client2.syncShutdownGracefully()
        client2 = self.createClient(identifier: "testPersistentSession_subscriber")
        _ = try client2.v5.connect(cleanStart: false).wait()
        // client2 should receive this publish as we have reconnected with clean session set to false
        try client.publish(to: "testPersistentAtLeastOnce", payload: payload, qos: .atLeastOnce).wait()
        Thread.sleep(forTimeInterval: 1)
        
        wait(for: [expectation], timeout: 5.0)
        
        try client.disconnect().wait()
        try client2.disconnect().wait()
    }
    
    func createClient(identifier: String, configuration: MQTTClient.Configuration = .init(version: .v5_0)) -> MQTTClient {
        MQTTClient(
            host: "localhost",
            port: 1883,
            identifier: identifier,
            eventLoopGroupProvider: .createNew,
            logger: nil,
            configuration: configuration
        )
    }
```

</details>

**Reply by adam-fowler** [reply]
You need to call `addPublishListener` again. The first publish listener was waiting for events on the old client

**Reply by ruicanas** [reply]
Isn't it possible to keep the old client or make replacements directly on it? Since I'm reconnecting and only want to change an authentication parameter, adding a publish listener once again seems like I'm using a sledgehammer to crack a nut.

**Reply by adam-fowler** [reply]
@ruicanas would this work for you #143

**Reply by ruicanas** [reply]
@adam-fowler It would and it works like a charm. Thanks for adding this extra parameter. It makes everything much smoother and more straightforward.

**Reply by adam-fowler** [reply]
Released in v2.10.0

**Comment by leisurehound** [comment]
> There are quite a few parameters to consider when doing a retry system.
> 
> * how long to wait between retries
> * Should wait time increase exponentially/linearly after each fail and by how much
> * Should wait time increase be jittered
> * Should it fail after so many retries
> * Should it fail after so much time
>   I would rather the individual wrote their own, instead of the client providing an implementation that tries to support every combination of the above

Is there a better mechanism to use than client.isActive() to determine if the connection is still valid? I'm still seeing .isActive() true when the network interface/connection has been brought down, not really allowing me to use it to impl my own retry mechanism.  

EDIT:  looks like my interval for checking isAlive was shorter than the keepAlive interval.  I was setting it near-ish the ping interval thinking a failed ping would result in it closing.  But the CloseListener is called around when the keepAlive interval expires

**Reply by adam-fowler** [reply]
The close listener is the best way to determine if the connection has been closed. This is called as soon as the connection is closed.

**Reply by adam-fowler** [reply]
Although it looks like we might not be reacting to all forms of network connection closure. I'll need to review this

**Reply by leisurehound** [reply]
I'm seeing the close listener anywhere from 45s to 90s after I disconnect from wifi.  However, in the console I see the `nw_proto_tcp_route_init [C1.1.1:2] no mtu received` message immediately.  would be great if we can see the call to the close listener and isActive() reflect the state sooner.

**Reply by adam-fowler** [reply]
Hi @leisurehound could you add an issue for this including all your platform details. 🙏

**Reply by leisurehound** [reply]
[done](https://github.com/swift-server-community/mqtt-nio/issues/145)

Separately, when the close listener is finally called, are all the client's eventloops terminated? On both linux and macOS?

**Reply by adam-fowler** [reply]
The eventLoopGroup is only shutdown when you call `MQTTClient.shutdown`.  If you provide your own EventLoopGroup then it is never shutdown as it is assumed you are using it for other processes
